### PR TITLE
ui: Show some basic information about found packages in the results main page

### PR DESCRIPTION
**Pull request by Etsija**
Please see the individual commits for details.

Screenshot from 2024-10-01 07-25-08

### CODE REVIEW

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx [hunk 1]
```diff
@@ -97,7 +109,34 @@ const RunComponent = () => {
...
+  }
+
+  const packagesTotal = packages.pagination.totalCount;
+
+  // Find deduplicated package types in packages data.
+  const detectedPackageManagers = [
+    ...new Set(
+      packages.data.map((item) => item.identifier?.type).filter(Boolean)
+    ),
+  ];
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: This likely doesn't work as the query is paginated, so `packages` only contains the first page with the default size.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx [hunk 2]
```diff
@@ -97,7 +109,34 @@ const RunComponent = () => {
...
+
+  console.log(detectedPackageManagers);
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:139]: Should be removed.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx [hunk 3]
```diff
@@ -265,8 +304,9 @@ const RunComponent = () => {
                   className={`h-4 w-4 ${getStatusFontColor(ortRun.jobs.analyzer?.status)}`}
                 />
               )}
-              value={ortRun.jobs.analyzer ? '-' : 'N/A'}
-              className='hover:bg-muted/50'
+              value={ortRun.jobs.analyzer ? packagesTotal : 'N/A'}
+              description={`from ${detectedPackageManagers.length} package managers (${detectedPackageManagers.join(`,`)})`}
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:308]: IMO this would look better with a badge for each package manager instead of a sentence.

**Review by mmurto — CHANGES_REQUESTED** [review]

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx (hunk 3 above)
Comment (sschuberth) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:308]: I think a sentence is fine for now. For monorepos with several package managers badges would probably take up too much screen space.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx (hunk 1 above)
Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: Should this rather be changed to describe that the list is of the package sources, or something else that would describe it better? I believe to type to just be [ORT Identifier's type](https://github.com/oss-review-toolkit/ort/blob/a03471c7c5981bb1950d76f708bd151ca196dfbd/model/src/main/kotlin/Identifier.kt#L32-L37), so data to do what you're suggesting is not available here. The same is probably true for NPM/Yarn/pnpm all using NPM packages.

Comment (sschuberth) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: > I believe to type to just be [ORT Identifier's type](https://github.com/oss-review-toolkit/ort/blob/a03471c7c5981bb1950d76f708bd151ca196dfbd/model/src/main/kotlin/Identifier.kt#L32-L37), so data to do what you're suggesting is not available here.

The information about whether an `Identifier` belongs to a package or project is not available in the `Identifier` itself. What the code *should* do is to collect the types of the identifiers of all root projects in an ORT result instead of iterating over packages.

Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: > > I believe to type to just be [ORT Identifier's type](https://github.com/oss-review-toolkit/ort/blob/a03471c7c5981bb1950d76f708bd151ca196dfbd/model/src/main/kotlin/Identifier.kt#L32-L37), so data to do what you're suggesting is not available here.
> 
> The information about whether an `Identifier` belongs to a package or project is not available in the `Identifier` itself. What the code _should_ do is to collect the types of the identifiers of all root projects in an ORT result instead of iterating over packages.

IMO there can be value for both, showing what package managers are used to manage the project and what type of dependencies the project uses.

Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: So maybe for this part it would make sense to reword the text to something like "100 packages from 2 _registries_ (Npm, Maven)". The package managers used by the project can be added here or somewhere else when the data is available, but I wouldn't hold this PR to add that.

Comment (sschuberth) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: > IMO there can be value for both

Maybe, but for practical matters I've always found it to be much more useful to know which package managers / project types are used instead of which ecosystems packages stem from.

> So maybe for this part it would make sense to reword the text

I agree that this initial implementation could "cheat" by adjusting the text to the current implementation instead of making the implementation match my more complex ask. However, I'd propose a different wording, and say "100 packages of 2 types (Maven, NPM) found". Oh, and the list of package types should be sorted alphabetically to be deterministic.

Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: > Oh, and the list of package types should be sorted alphabetically to be deterministic.

Sorted by package count per type could be good and also deterministic.

Comment (sschuberth) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: Agreed, but I'd also be fine with not having per-type counts in this first iteration. At least not in the card; that's probably too detailed information that could go to a header on the "Packages" tab.

Comment (mnonnenmacher) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: > "100 packages of 2 types (Maven, NPM) found"

How about "100 packages from 2 ecosystems (Maven, NPM) found" instead?

Comment (sschuberth) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:137]: I was at first thinking about almost exactly the same as I was using "ecosystems" myself [above](https://github.com/eclipse-apoapsis/ort-server/pull/1125#discussion_r1781198645), but then I didn't want to coin yet another new term. On the other hand, we should have the liberty to use "new" terms in the UI if we believe the (non-ORT) user is familiar with them.

TL;DR  "100 packages from 2 ecosystems (Maven, NPM)" would be ok with me.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx [hunk 4]
```diff
@@ -97,7 +118,32 @@ const RunComponent = () => {
...
+  }
+
+  const packagesTotal = packages.pagination.totalCount;
+
+  // Find de-duplicated package types in packages data and sort alphabetically.
+  const ecoSystems = [
+    ...new Set(
+      packages.data.map((item) => item.identifier?.type).filter(Boolean)
+    ),
+  ].sort();
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/index.tsx:146]: This should maybe be memoized as the array can be quite large.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/-components/packages-statistics-card.tsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        ...new Set(
+          packages?.data.map((item) => item.identifier?.type).filter(Boolean)
+        ),
+      ].sort(),
+    [packages?.data]
+  );
+
+  if (packagesIsPending) {
+    return <LoadingIndicator />;
+  }
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/-components/packages-statistics-card.tsx:71]: This would look better if the card including the title and icon etc was there while the data is pending. I would maybe try [skeleton](https://ui.shadcn.com/docs/components/skeleton) at first.

**Review by mmurto — CHANGES_REQUESTED** [review]

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/-components/packages-statistics-card.tsx (hunk 5 above)
Comment (mmurto) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/-components/packages-statistics-card.tsx:71]: Or just a loading replacing the package count, skeleton doesn't look too good on first try.

Comment (Etsija) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/-components/packages-statistics-card.tsx]: I replaced all "real" data with a loading indicator, because there's no point in showing "<spinner> from ??? ecosystems (???)". So during loading and processing, only the title and icon are shown.

File: ui/src/components/statistics-card.tsx [hunk 6]
```diff
@@ -44,8 +46,11 @@ export const StatisticsCard = ({
         </CardTitle>
       </CardHeader>
       <CardContent className='text-sm'>
-        <div className='flex items-center justify-between'>
-          <div className='text-2xl font-bold'>{value ? value : '-'}</div>
+        <div className='flex flex-col'>
+          <div className='text-2xl font-bold'>
+            {value ? value === 'loading' ? <LoadingIndicator /> : value : '-'}
+          </div>
```
Comment (mmurto) [inline ui/src/components/statistics-card.tsx:52]: This should not be based on matching a magic string like "loading". The simplest solution could be to type value as `React.ReactNode`.

**Review by mmurto — CHANGES_REQUESTED** [review]

**Comment by mmurto** [comment]
There may be a bug in the packages endpoint - please hold on merging this until I've verified it.

**Comment by Etsija** [comment]
> There may be a bug in the packages endpoint - please hold on merging this until I've verified it.

I disabled auto-merge.

**Comment by sschuberth** [comment]
> > There may be a bug in the packages endpoint - please hold on merging this until I've verified it.
> 
> I disabled auto-merge.

So, this is about https://github.com/eclipse-apoapsis/ort-server/issues/1137, right? But does that issue actually block the merge of this PR?

**Comment by sschuberth** [comment]
> If this is merged before fixing #1137, the UI will call the endpoint and show some errors to the end users.

Now that #1137 is fixed, can we proceed with this?

**Comment by mmurto** [comment]
> > If this is merged before fixing #1137, the UI will call the endpoint and show some errors to the end users.
> 
> Now that #1137 is fixed, can we proceed with this?

No objections from me.

**Comment by sschuberth** [comment]
> No objections from me.

So, mind re-reviewing to resolve your "1 change requested" status, @mmurto?

**Review by mmurto — DISMISSED** [review]

**Comment by sschuberth** [comment]
Looks like there are build errors now. @Etsija can you please rebase and investigate?

**Review by mmurto — APPROVED** [review]
